### PR TITLE
[Perf][Attention] Reduce GLM sparse MLA preparation overhead

### PR DESCRIPTION
**Pull request by GirasoleY**
## Purpose

Reduce query packing and sparse-index preparation overhead around FlashInfer NoPE MLA attention.

The three commits:

1. Reuse a contiguous NoPE query when its rotary tail is empty, avoiding a redundant `torch.cat` before FP8 quantization. Preserve concatenation for nonempty rotary dimensions and strided queries.
2. Use one padded remap program per query for counted sparse capacities up to 4096. Initialize padding and reduce valid counts inside the remap kernel, removing separate fills and atomic accumulation. Preserve larger tiled cases and reusable buffers. Name the device entry point `sparse_mla_index_remap_kernel` for trace attribution.
3. Simplify empty-query handling: fuse dummy-entry preparation and safe-length calculation while preserving raw counts; skip preparation and masking for resident TP; mask only LSE for DCP and let its existing merge suppress zero-weight outputs, including NaNs. Preserve PyTorch output masking for HiSparse without LSE. There is no custom output-zeroing kernel.

Duplicate-work checks found no matching open PR. Related PRs #54258 (separate ROCm remap kernel), #38354 (DeepSeek pre-attention GEMM/RoPE fusion), and #57161 (GLM kpool/indexer scheduling) address different kernels or stages. No existing issue is being claimed or closed.

## Correctness

Prior TP validation on the experiment stack used GLM-5.3-Flash, TP8+EP8, all custom ops enabled, eager execution, no prefix cache, and exactly 32768 input tokens with either 1 or 16 generated tokens. Across 12 repeatability requests, all 8976 sparse-layer calls across eight ranks had positive valid counts (minimum 1). A direct full-shape FP8 attention comparison at 32768 queries × 8 heads × 512 dimensions, capacity 2176, matched the previous implementation bit for bit.

## Performance

Controlled end-to-end comparison on four GB300 GPUs, TP4+EP4, eager, `custom_ops=["all"]`, no prefix cache, no speculative decoding, and FP8 E4M3 KV cache. The actual upstream PR model stayed loaded while an idle-worker extension switched only the three preparation changes to their PR-base implementations. All four ranks verified the implementation selection. Compatible prebuilt extensions were reused.

Both implementations were warmed, then measured in four alternating baseline/PR and PR/baseline rounds with five matched samples per workload per round (20 pairs each). Each request had exactly 32768 input tokens, with a 32768-token prefill scheduling limit. Prefill generated one token; decode generated 128 tokens. These are client-observed serving latencies.

| Metric | PR base | PR | Latency reduction |
| --- | ---: | ---: | ---: |
| 32K prefill TTFT, median | 518.25 ms | 507.61 ms | 2.05% |
| Decode inter-token latency, median | 122.85 ms/token | 120.77 ms/token | 1.70% |

All 20 pairs improved for each workload. Per-round median reductions ranged from 1.78–2.27% for prefill and 1.21–1.98% for decode. These measurements apply to eager TP4 execution; CUDA-graph decode and other parallel configurations were not measured in this comparison.

Earlier isolated measurements on the same GB300 GPU, using CUDA graph replay and the median of five samples with 40 repetitions/sample:

| Operation | Before | Optimized |
| --- | ---: | ---: |
| Remap including initialization | 531.90 µs | 282.93 µs |
| Safe lengths and dummy entry | 12.32 µs | 4.12 µs |

Shape: 32768 queries × 8 heads × 512 dimensions, capacity 2176. These are preparation microbenchmarks from the experiment stack. Resident TP now skips safe-length preparation entirely. Earlier tracing also confirmed removal of all 11 redundant query concatenations. The final upstream revision has not been reprofiled; the controlled end-to-end results above are separate from these older microbenchmarks. Measurements of the removed output-zeroing kernel are excluded.

### CODE REVIEW

File: vllm/v1/attention/backends/mla/sparse_utils.py [hunk 1]
```diff
@@ -685,3 +696,46 @@ def triton_filter_and_convert_dcp_index(
...
 
 _CONVERT_REQ_INDEX_TO_GLOBAL_INDEX_KERNEL = ConvertReqIndexToGlobalIndexKernel()
+
+
+@triton.jit
+def sparse_mla_prepare_safe_lengths_kernel(
+    indices_ptr,
+    counts_ptr,
+    safe_lengths_ptr,
+    NUM_TOKENS: tl.constexpr,
```
Comment (ZJY0516) [inline vllm/v1/attention/backends/mla/sparse_utils.py]: This will trigger JIT recompilation

**Review by ZJY0516 — APPROVED** [review]
LGTM

File: vllm/v1/attention/backends/mla/sparse_utils.py (hunk 1 above)
Comment (GirasoleY) [reply vllm/v1/attention/backends/mla/sparse_utils.py]: Addessed in https://github.com/vllm-project/vllm/pull/57458/changes#diff-cf5fac6184147054a98024c70b8b09e414dabd6aa1c5d1a40d124027f1d0e5f6R701
